### ISSUE TITLE
Make child nodes export variables be editable in the scene root inspector

### ISSUE THREAD

**Issue by TushieSushies** [godotengine/godot-proposals#7803]
### Describe the project you are working on

A 2.5D RPG where the player can interact with objects and play dialogue by interacting with npc's / objects

### Describe the problem or limitation you are having in your project

If I instance a scene, and I want to access or edit an exported variable of a child node in the inspector, currently the only way is to make a setget variable in the root object, this is not ideal and is prone to mistakes.
We want to edit out what line of text/asset is played when interacting and we want to keep the project structure consistent with Godot's node approach, such as keeping the dialogue logic in a dialogue component, but editing what asset is used in the variable in the dialogue component is not available on instanced scenes.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

If child node exported variables appeared, we wouldn't need to use setget functions and it streamline the work greatly.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Say we have a component that plays dialogue, we called it DialogueComponent, and is a child of the node TestCube:
![image](https://github.com/godotengine/godot-proposals/assets/145011115/7c364f08-1932-4dd7-bed1-cf3251cab0e8)

The DialogueComponent uses an Exported variable to know what asset to play when ordered from a function inside the node.
![image](https://github.com/godotengine/godot-proposals/assets/145011115/e413c989-743b-4548-912f-17b0f1c5c13e)

For exported variables, they could be visible to the root node like in this mockup here:
![Untitled-2](https://github.com/godotengine/godot-proposals/assets/145011115/05740839-29db-4ee3-800f-d3a30401cd33)

Resulting in them being accessible in instanced scenes like this where they otherwise wouldn't be:
![Untitled-3](https://github.com/godotengine/godot-proposals/assets/145011115/c8fdb025-f74f-4594-a279-4f76325e70c3)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes, with a setget function in the root node script, however it's prone to typos and null variables crashing the game

### Is there a reason why this should be core and not an add-on in the asset library?

This will greatly improve the flow of instancing scenes where editing exported variables of child nodes is required

**Comment by Zireael07** [comment]
Counterpoint: I have scenes where child nodes have a TON of exported variables. Exposing them all on the root would be UNMANAGEABLE

I use editable scenes instead

**Comment by TushieSushies** [comment]
Then I think it could be another type of variable attribute if not exported

**Comment by jmarceno** [comment]
Exported variables are "attached" to the script attached to the node.
This proposal as it is, would break the concept as a recursive scan would need to done in every script attached to the root node via its children. I do not know the inners of how the script association works, but I can see this being very prone to break, and going against this core concept (1 node 1 script).

Besides this, as @Zireael07 said, I can see a good number of cases where users would not want this to happen, as the root node of complex scenes would become unmanageable.

Doing what you being doing (setting it via a var at the root node with a setter), is completely fine in my opinion. Just check for null values before setting the value and you should be fine.

**Comment by 3lis4** [comment]
I think this proposal might be interesting as a **syntactic sugar** to limit the amount of "useless" setters when many exported variables of the child nodes need to be exported as well on the parent node.

It could be an annotation exposing a child node's exported variable. Something like `@export_child(path_to_child_node:var_name)` in parent node's script.

In our project we have many of these setters and such an annotation would greatly improve readability and be easier/faster to write. Also, auto-completion could help to write the child node's path (on the same way it is for $ syntax).

**Comment by TushieSushies** [comment]
I agree, @export might not be a good idea due to the points made by other users, but another way to export to root node will surely come in handy instead of relying on setget

**Comment by ywmaa** [comment]
I really need this feature too, but please the idea of exporting everything manually to the root is not good.

> It could be an annotation exposing a child node's exported variable. Something like @export_child(path_to_child_node:var_name) in parent node's script.

This sounds much better.

**Comment by TushieSushies** [comment]
I vote their for their proposal as well!
Now we just need someone kind enough to implement it ^_^

**Comment by ywmaa** [comment]
I may try to do it myself, but I guess we need to have full understanding of what it will do, because I will already try tweaking GD script which is very new to me, so I don't wan't to also go in it blind.

`@export_child(path_to_child_node:var_name)`
this works as both NodePath and @export_category ?

so it will create a new category in the root of the node containing the name/class of the exported child with its exported categories too.

 do we need custom naming ?
 should it be in the root, or at the top like normal @export ?
 what happens when you set the scene to editable, and edit the children values, will they reflect to the parent ?
 what happens when we use this recursively (child also exported another child) ?

**Comment by ywmaa** [comment]
Here we go, I made a draft PR, you can test it, and/or provide feedback.

**Comment by AttackButton** [comment]
As @Zireael07 said, what could this do that `editable_child` couldn't do better?

**Comment by AttackButton** [comment]
There is already a discussion about this enhancement in the `editable_children`: https://github.com/godotengine/godot-proposals/issues/3248.

Furthermore, there are two prs (apparently incomplete) trying to implement that proposal: https://github.com/godotengine/godot/pull/63307 and https://github.com/godotengine/godot/pull/60974.

I think that the ideal would be to implement a better version than these prs for `editable_children`, perhaps creating an annotation or property allowing editable_children selecting which resources to expose.

**Comment by ywmaa** [comment]
@AttackButton 

I am not sure, but did you look at my PR ?

Since you are mentioning an annotation ?
Is the @export_child annotation not what you mean ?

**Comment by AttackButton** [comment]
I mentioned an annotation or a property to enable `editable_children` by default for the instantiated scenes from a given base scene.

The only question about this proposal is whether it is not already covered by what `editable_children` does.

**Comment by TushieSushies** [comment]
I agree the features in the proposal https://github.com/godotengine/godot-proposals/issues/3248 should be implemented instead, it’s a more elegant and less clumsy implementation of my idea

**Comment by ywmaa** [comment]
Yeah it looks better to me, too.

Is there any PR for it ?

**Comment by Shadowblitz16** [comment]
Why not treat scene roots as packages instead of the root node?
The scene package by default could have nothing exported and user's could choose what properties on what node's are exported.

**Comment by MrModez** [comment]
Crazy that we still don't have any solution for this. It's nearly impossible to do level design and the workarounds feel extremely hacky..

**Comment by AThousandShips** [comment]
@MrModez Please don't bump without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by EricDeng1001** [comment]
FYI， in C#， I write custom Source Generator to generator code to allow me edit child node properties in root node.

**Comment by Milo-o-Fun** [comment]
I really need this function. In fact, it doesn't have to export all the child nodes directly. Just exporting some key ones is fine. It's a bit like "favorite property". I think we can do it this way:
@export_child_node var child = %"Child" ## this will make followers become a Group.
@export_property(child.size) var child_size :Vector2D

**Comment by Xesa** [comment]
I've made a plugin that does exactly this. You can choose which properties will get exposed and which nodes will show these exposed properties. So you don't end up with a mess in your editor.

Would be great to have this natively in the engine, but in the meantime...

https://github.com/Xesa/expose-them-properties
